### PR TITLE
ACM-2940 Setup cluster-manager webhook proxying in hosted-mode MCE

### PR DESCRIPTION
**Pull request by cameronmwall**
The PR create several new resources when enabling cluster manager so as to allow proxying for the web hooks. The most important of these is the konnectivity deployment which takes in the addresses of the two web hook services. These addresses are also added to be cluster manager itself. When this deployment is up and running in the MCE target namespace, proxying should be working. In order to test this I get a oc api-resources on the hosted cluster which was expected to complete successfully

Issue: https://issues.redhat.com/browse/ACM-2940

### CODE REVIEW

File: controllers/hosted.go [hunk 1]
```diff
@@ -366,21 +366,117 @@ func (r *MultiClusterEngineReconciler) ensureHostedClusterManager(ctx context.Co
...
+		err = r.Client.Patch(ctx, cmTemplate, client.Apply, &client.PatchOptions{Force: &force, FieldManager: "backplane-operator"})
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error applying object Name: %s Kind: %s, %w", cmTemplate.GetName(), cmTemplate.GetKind(), err)
+		}
+		log.Info(fmt.Sprintf("Re-reconciling to get proxy addresses"))
+		return ctrl.Result{Requeue: true}, nil
+	} else {
+
+		// Apply clustermanager
+		cmTemplate := foundation.HostedClusterManager(mce, r.Images, registrationWebhookService.Spec.ClusterIP, workWebhookService.Spec.ClusterIP)
```
Comment (JakobGray) [inline controllers/hosted.go:447]: Is it possible MCE will declare itself available before the clusterIPs are set?

File: pkg/templates/charts/hosting/konnectivity/templates/cm-webhook-agent.yaml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        - --probe-interval
+        - 30s
+        - --sync-interval
+        - 1m
+        - --sync-interval-cap
+        - 5m
+        command:
+        - /usr/bin/proxy-agent
+        # MCE has an external image entry for the proxy agent image.  SHould we use that?
+        image: quay.io/stolostron/apiserver-network-proxy:latest
```
Comment (JakobGray) [inline pkg/templates/charts/hosting/konnectivity/templates/cm-webhook-agent.yaml:70]: Is there an image that needs to be onboarded for this deployment?

**Review by JakobGray — COMMENTED** [review]
It would be helpful if you could add a description of the changes this PR introduces, and maybe also what to look for to know that it's doing what it should.

File: pkg/utils/annotations.go [hunk 3]
```diff
@@ -111,3 +115,42 @@ func GetHostedCredentialsSecret(mce *backplanev1.MultiClusterEngine) (types.Name
 	}
 	return nn, nil
 }
+
+func GetKonnectivitySecret(mce *backplanev1.MultiClusterEngine) (types.NamespacedName, error) {
+	nn := types.NamespacedName{}
+	if mce.Annotations == nil || mce.Annotations[AnnotationHyperShiftNamespace] == "" {
+		return nn, fmt.Errorf("no kubeconfig secret annotation defined in %s", mce.Name)
```
Comment (JakobGray) [inline pkg/utils/annotations.go:122]: I think this is a typo

**Review by JakobGray — COMMENTED** [review]
On initial run with these changes it didn't seem to change much, until I found out I had to poke the hosted MCE to get it to progress. Despite that the MCE was showing as Available. I believe it's because this doesn't meet the criteria:
> MCE status tracks success of resources applied

So MCE is still only showing one Deployment and one ClusterManager in its status.

Can you provide a list of commands to run to verify everything is deployed as it should be? For example:
```
oc get deploy cluster-manager-webhook-konnectivity-agent -n <mceNS>
oc get service cluster-manager-work-webhook -n <mceNS>
KUBECONFIG=<hosted-kubeconfig> oc get service .....
```

I tried the oc api-resources call and nothing stood out to me before or after the changes were applied so I'm not sure if anything happened.

**Review by eemurphy — COMMENTED** [review]
I have no other comments than what Jakob has added, otherwise was able to do oc api-resources and see the konnectivity deployment as expected

**Pull request closed without merging**